janino/src/main/java/org/codehaus/janino/UnitCompiler.java

```diff
@@ -9324,35 +9326,35 @@
 
         if (maximallySpecificIInvocables.size() == 1) return (IInvocable) maximallySpecificIInvocables.get(0);
 
-        ONE_NON_ABSTRACT_INVOCABLE:
+        ONE_CONCRETE_INVOCABLE:
         if (maximallySpecificIInvocables.size() > 1 && iInvocables[0] instanceof IClass.IMethod) {
 
             // Check if all methods have the same signature (i.e. the types of all their parameters are identical) and
-            // exactly one of the methods is non-abstract (JLS7 15.12.2.2.BL2.B1).
-            IClass.IMethod theNonAbstractMethod = null;
+            // exactly one of the methods is concrete (JLS8 15.12.2.5.B8.B1).
+            IClass.IMethod theConcreteMethod = null;
             {
                 Iterator<IClass.IInvocable> it                          = maximallySpecificIInvocables.iterator();
                 IClass.IMethod              m                           = (IClass.IMethod) it.next();
                 final IClass[]              parameterTypesOfFirstMethod = m.getParameterTypes();
                 for (;;) {
-                    if (!m.isAbstract()) {
-                        if (theNonAbstractMethod == null) {
-                            theNonAbstractMethod = m;
+                    if (!m.isAbstract() && !m.isDefault()) {

```

Comment:
This doesn't look right to me. Inlining `isDefault`, you get `!m.isAbstract() && (m.isAbstract() || !m.method.getDeclaringClass().isInterface();)`, which is equivalent to `!m.isAbstract() && !m.method.getDeclaringClass().isInterface();`.

I think this condition was right, because "concrete (that is, non-abstract or default)" can actually just be tested as `!m.isAbstract()`. My interpretation is that the error condition below needs to be relaxed, and allow for multiple non-abstract methods, as long as they are all defined in interfaces:

> Otherwise, if all the maximally specific methods are abstract or default,

which expands to:

> Otherwise, if all the maximally specific methods are abstract or (non-abstract and defined in an interface),